File: publish-spywizards-articles.py
#!/usr/bin/env python3
"""
SpyWizards Article Publisher via WordPress REST API
Publishes articles from markdown files to WordPress with Rank Math SEO metadata
"""

import requests
import base64
import os
import re
from pathlib import Path

# Configuration
SITE_URL = "https://spywizards.com/blog/wp-json"
USERNAME = "Aiwisemind"
PASSWORD = "IbGx RiYc qzoX swu5 nZtb mR37"

# Content folder
CONTENT_FOLDER = Path(__file__).parent / "spywizards-content"

# Authentication
auth = base64.b64encode(f"{USERNAME}:{PASSWORD}".encode()).decode()
headers = {
    "Authorization": f"Basic {auth}",
    "Content-Type": "application/json"
}

def extract_seo_frontmatter(content):
    """Extract SEO metadata from markdown frontmatter-style headers"""
    seo = {
        "title": "",
        "description": "",
        "focus_keyword": "",
        "categories": ["Cybersecurity", "Social Media Security"],
        "tags": []
    }
    
    # Look for title patterns
    title_match = re.search(r'^# (.+)$', content, re.MULTILINE)
    if title_match:
        seo["title"] = title_match.group(1).strip()
    
    # Look for focus keyword
    keyword_match = re.search(r'\*\*Focus Keyword[:\-]\*\*\s*(.+)', content)
    if keyword_match:
        seo["focus_keyword"] = keyword_match.group(1).strip()
    
    # Look for meta description pattern
    desc_match = re.search(r'\*\*Meta Description[:\-]\*\*\s*(.+)', content)
    if desc_match:
        seo["description"] = desc_match.group(1).strip()
    
    # Extract tags from last lines
    tags_match = re.search(r'Tags[:\-]\s*(.+)', content)
    if tags_match:
        seo["tags"] = [t.strip() for t in tags_match.group(1).split(',')]
    
    # Look for categories
    cat_match = re.search(r'Categories[:\-]\s*(.+)', content)
    if cat_match:
        seo["categories"] = [c.strip() for c in cat_match.group(1).split(',')]
    
    return seo

def convert_markdown_to_html(content):
    """Simple markdown to HTML conversion"""
    # Headers
    content = re.sub(r'^### (.+)$', r'<h3>\1</h3>', content, flags=re.MULTILINE)
    content = re.sub(r'^## (.+)$', r'<h2>\1</h2>', content, flags=re.MULTILINE)
    content = re.sub(r'^# (.+)$', r'<h1>\1</h1>', content, flags=re.MULTILINE)
    
    # Bold
    content = re.sub(r'\*\*([^*]+)\*\*', r'<strong>\1</strong>', content)
    
    # Paragraphs (split by double newlines)
    paragraphs = content.split('\n\n')
    html_paragraphs = []
    for p in paragraphs:
        if p.strip() and not p.startswith('<h'):
            html_paragraphs.append(f"<p>{p.strip()}</p>")
        elif p.strip():
            html_paragraphs.append(p.strip())
    
    return '\n'.join(html_paragraphs)

def get_category_id(category_name):
    """Get WordPress category ID by name"""
    response = requests.get(
        f"{SITE_URL}/wp/v2/categories",
        headers=headers,
        params={"search": category_name}
    )
    if response.status_code == 200:
        categories = response.json()
        for cat in categories:
            if cat["name"].lower() == category_name.lower():
                return cat["id"]
    return None

def get_tag_id(tag_name):
    """Get WordPress tag ID by name"""
    response = requests.get(
        f"{SITE_URL}/wp/v2/tags",
        headers=headers,
        params={"search": tag_name}
    )
    if response.status_code == 200:
        tags = response.json()
        for tag in tags:
            if tag["name"].lower() == tag_name.lower():
                return tag["id"]
    return None

def create_category(category_name):
    """Create a new category and return its ID"""
    response = requests.post(
        f"{SITE_URL}/wp/v2/categories",
        headers=headers,
        json={"name": category_name}
    )
    if response.status_code == 201:
        return response.json()["id"]
    return None

def create_tag(tag_name):
    """Create a new tag and return its ID"""
    response = requests.post(
        f"{SITE_URL}/wp/v2/tags",
        headers=headers,
        json={"name": tag_name}
    )
    if response.status_code == 201:
        return response.json()["id"]
    return None

def publish_article(filepath):
    """Publish a single article to WordPress"""
    print(f"\n📄 Publishing: {filepath.name}")
    
    # Read markdown file
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Extract SEO metadata
    seo = extract_seo_frontmatter(content)
    
    # Convert to HTML (simple conversion)
    html_content = convert_markdown_to_html(content)
    
    # Create slug from filename
    slug = filepath.stem.lower().replace('_', '-')
    
    # Prepare post data
    post_data = {
        "title": seo["title"] or filepath.stem.replace('-', ' ').title(),
        "content": html_content,
        "slug": slug,
        "status": "publish"
    }
    
    # Create categories and tags
    category_ids = []
    for cat_name in seo["categories"]:
        cat_id = get_category_id(cat_name)
        if not cat_id:
            cat_id = create_category(cat_name)
        if cat_id:
            category_ids.append(cat_id)
    
    tag_ids = []
    for tag_name in seo["tags"]:
        tag_id = get_tag_id(tag_name)
        if not tag_id:
            tag_id = create_tag(tag_name)
        if tag_id:
            tag_ids.append(tag_id)
    
    if category_ids:
        post_data["categories"] = category_ids
    if tag_ids:
        post_data["tags"] = tag_ids
    
    # Add Rank Math SEO metadata (as custom fields via REST API or meta schema)
    # Note: Rank Math stores SEO data in post meta
    post_data["meta"] = {
        "rank_math_title": seo["title"],
        "rank_math_description": seo["description"],
        "rank_math_focus_keyword": seo["focus_keyword"]
    }
    
    # Publish post
    response = requests.post(
        f"{SITE_URL}/wp/v2/posts",
        headers=headers,
        json=post_data
    )
    
    if response.status_code == 201:
        post = response.json()
        print(f"✅ Published: {post['link']}")
        return {"success": True, "url": post['link'], "id": post['id']}
    else:
        print(f"❌ Failed: {response.status_code} - {response.text[:200]}")
        return {"success": False, "error": response.text}

def main():
    """Main function to publish all articles"""
    print("🚀 Starting SpyWizards Article Publisher")
    print(f"📁 Content folder: {CONTENT_FOLDER}")
    print(f"🌐 Site: {SITE_URL}")
    
    # Find all markdown files
    md_files = sorted(CONTENT_FOLDER.glob("*.md"))
    
    print(f"\n📊 Found {len(md_files)} articles to publish\n")
    
    results = []
    for md_file in md_files:
        result = publish_article(md_file)
        results.append(result)
    
    # Summary
    print("\n" + "="*50)
    print("📈 PUBLICATION SUMMARY")
    print("="*50)
    successful = sum(1 for r in results if r.get("success"))
    failed = len(results) - successful
    print(f"✅ Successful: {successful}")
    print(f"❌ Failed: {failed}")
    
    if successful > 0:
        print("\n🔗 Published URLs:")
        for r in results:
            if r.get("success"):
                print(f"  - {r['url']}")
    
    # Save log
    log_file = CONTENT_FOLDER.parent / "spywizards-posts-log.md"
    with open(log_file, 'w', encoding='utf-8') as f:
        f.write("# SpyWizards.com Posts Log\n\n")
        f.write(f"**Published:** {successful}\n")
        f.write(f"**Failed:** {failed}\n\n")
        f.write("## Published Posts\n\n")
        for r in results:
            if r.get("success"):
                f.write(f"- [{r['url']}]({r['url']}) - ID: {r['id']}\n")
    print(f"\n📝 Log saved to: {log_file}")

if __name__ == "__main__":
    main()

File: publish-spywizards-articles.js
/**
 * SpyWizards Article Publisher via WordPress REST API
 * Publishes articles from markdown files to WordPress with Rank Math SEO metadata
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// Configuration
const SITE_URL = 'https://spywizards.com/blog/wp-json';
const USERNAME = 'Aiwisemind';
const PASSWORD = 'IbGx RiYc qzoX swu5 nZtb mR37';
const CONTENT_FOLDER = path.join(__dirname, 'spywizards-content');

// Auth header
const auth = Buffer.from(`${USERNAME}:${PASSWORD}`).toString('base64');
const headers = {
    'Authorization': `Basic ${auth}`,
    'Content-Type': 'application/json'
};

function extractSeoFrontmatter(content) {
    const seo = {
        title: '',
        description: '',
        focusKeyword: '',
        categories: ['Cybersecurity', 'Social Media Security'],
        tags: []
    };
    
    // Title
    const titleMatch = content.match(/^# (.+)$/m);
    if (titleMatch) seo.title = titleMatch[1].trim();
    
    // Focus keyword
    const keywordMatch = content.match(/\*\*Focus Keyword[:\-]\*\*\s*(.+)/i);
    if (keywordMatch) seo.focusKeyword = keywordMatch[1].trim();
    
    // Meta description
    const descMatch = content.match(/\*\*Meta Description[:\-]\*\*\s*(.+)/i);
    if (descMatch) seo.description = descMatch[1].trim();
    
    // Tags
    const tagsMatch = content.match(/Tags[:\-]\s*(.+)/i);
    if (tagsMatch) {
        seo.tags = tagsMatch[1].split(',').map(t => t.trim());
    }
    
    // Categories
    const catMatch = content.match(/Categories[:\-]\s*(.+)/i);
    if (catMatch) {
        seo.categories = catMatch[1].split(',').map(c => c.trim());
    }
    
    return seo;
}

function convertMarkdownToHtml(content) {
    let html = content;
    
    // Headers
    html = html.replace(/^### (.+)$/gm, '<h3>$1</h3>');
    html = html.replace(/^## (.+)$/gm, '<h2>$1</h2>');
    html = html.replace(/^# (.+)$/gm, '<h1>$1</h1>');
    
    // Bold
    html = html.replace(/\*\*([^*]+)\*\*/g, '<strong>$1</strong>');
    
    // Paragraphs
    const paragraphs = html.split(/\n\n+/);
    const htmlParagraphs = paragraphs.map(p => {
        p = p.trim();
        if (p && !p.startsWith('<')) {
            return `<p>${p}</p>`;
        }
        return p;
    });
    
    return htmlParagraphs.join('\n');
}

function httpRequest(url, method = 'GET', data = null) {
    return new Promise((resolve, reject) => {
        const urlObj = new URL(url);
        const options = {
            hostname: urlObj.hostname,
            port: 443,
            path: urlObj.pathname + urlObj.search,
            method: method,
            headers: headers
        };
        
        const req = https.request(options, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve({ status: res.statusCode, data: JSON.parse(body) });
                } catch (e) {
                    resolve({ status: res.statusCode, data: body });
                }
            });
        });
        
        req.on('error', reject);
        req.setTimeout(30000, () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
        
        if (data) {
            req.write(JSON.stringify(data));
        }
        req.end();
    });
}

async function getCategoryId(categoryName) {
    const result = await httpRequest(`${SITE_URL}/wp/v2/categories?search=${encodeURIComponent(categoryName)}`);
    if (result.data && result.data.length > 0) {
        const cat = result.data.find(c => c.name.toLowerCase() === categoryName.toLowerCase());
        if (cat) return cat.id;
    }
    return null;
}

async function createCategory(categoryName) {
    const result = await httpRequest(`${SITE_URL}/wp/v2/categories`, 'POST', { name: categoryName });
    if (result.status === 201) return result.data.id;
    return null;
}

async function getTagId(tagName) {
    const result = await httpRequest(`${SITE_URL}/wp/v2/tags?search=${encodeURIComponent(tagName)}`);
    if (result.data && result.data.length > 0) {
        const tag = result.data.find(t => t.name.toLowerCase() === tagName.toLowerCase());
        if (tag) return tag.id;
    }
    return null;
}

async function createTag(tagName) {
    const result = await httpRequest(`${SITE_URL}/wp/v2/tags`, 'POST', { name: tagName });
    if (result.status === 201) return result.data.id;
    return null;
}

async function publishArticle(filepath) {
    console.log(`\n📄 Publishing: ${path.basename(filepath)}`);
    
    const content = fs.readFileSync(filepath, 'utf8');
    const seo = extractSeoFrontmatter(content);
    const htmlContent = convertMarkdownToHtml(content);
    
    const slug = path.basename(filepath, '.md').toLowerCase().replace(/_/g, '-');
    
    // Get category IDs
    const categoryIds = [];
    for (const catName of seo.categories) {
        let catId = await getCategoryId(catName);
        if (!catId) catId = await createCategory(catName);
        if (catId) categoryIds.push(catId);
    }
    
    // Get tag IDs
    const tagIds = [];
    for (const tagName of seo.tags) {
        let tagId = await getTagId(tagName);
        if (!tagId) tagId = await createTag(tagName);
        if (tagId) tagIds.push(tagId);
    }
    
    const postData = {
        title: seo.title || path.basename(filepath, '.md').replace(/-/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
        content: htmlContent,
        slug: slug,
        status: 'publish'
    };
    
    if (categoryIds.length > 0) postData.categories = categoryIds;
    if (tagIds.length > 0) postData.tags = tagIds;
    
    // Rank Math SEO meta
    postData.meta = {
        rank_math_title: seo.title,
        rank_math_description: seo.description,
        rank_math_focus_keyword: seo.focusKeyword
    };
    
    const result = await httpRequest(`${SITE_URL}/wp/v2/posts`, 'POST', postData);
    
    if (result.status === 201) {
        console.log(`✅ Published: ${result.data.link}`);
        return { success: true, url: result.data.link, id: result.data.id };
    } else {
        console.log(`❌ Failed: ${result.status}`);
        return { success: false, error: result.data };
    }
}

async function main() {
    console.log('🚀 Starting SpyWizards Article Publisher');
    console.log(`📁 Content folder: ${CONTENT_FOLDER}`);
    console.log(`🌐 Site: ${SITE_URL}`);
    
    // Find all markdown files
    const files = fs.readdirSync(CONTENT_FOLDER).filter(f => f.endsWith('.md')).sort();
    
    console.log(`\n📊 Found ${files.length} articles to publish\n`);
    
    const results = [];
    for (const file of files) {
        const result = await publishArticle(path.join(CONTENT_FOLDER, file));
        results.push(result);
    }
    
    // Summary
    console.log('\n' + '='.repeat(50));
    console.log('📈 PUBLICATION SUMMARY');
    console.log('='.repeat(50));
    const successful = results.filter(r => r.success).length;
    const failed = results.length - successful;
    console.log(`✅ Successful: ${successful}`);
    console.log(`❌ Failed: ${failed}`);
    
    if (successful > 0) {
        console.log('\n🔗 Published URLs:');
        for (const r of results) {
            if (r.success) console.log(`  - ${r.url}`);
        }
    }
    
    // Save log
    const logFile = path.join(CONTENT_FOLDER, '..', 'spywizards-posts-log.md');
    let logContent = '# SpyWizards.com Posts Log\n\n';
    logContent += `**Published:** ${successful}\n`;
    logContent += `**Failed:** ${failed}\n\n`;
    logContent += '## Published Posts\n\n';
    for (const r of results) {
        if (r.success) {
            logContent += `- [${r.url}](${r.url}) - ID: ${r.id}\n`;
        }
    }
    fs.writeFileSync(logFile, logContent);
    console.log(`\n📝 Log saved to: ${logFile}`);
}

main().catch(console.error);
